add TwoQubitNoiseModel and tests

DuckTigger opened the pull request:
Adds a two qubit noise model. This applies with either weighted or equal probability the combinations of I, X, Y, and Z gates after a two qubit gate has been applied. This is so we can set a two qubit gate to have a higher probability of noise, e.g. https://arxiv.org/pdf/quant-ph/0410199 pg.4 Error Model and Assumptions.

New gates were added to allow for combinations such as the IX or YX gate - this may be overkill.

Testing was added to the test file to 

Is used in the simulator like: 
```
simulator = cirq.DensityMatrixSimulator(noise=TwoQubitNoiseModel(cirq.depolarize(4*self.noise_prob / 5),
                                                                             two_qubit_depolarize(self.noise_prob)))
```
Let me know if this is not required.

dabacon commented:
Could your run
`./check/format-incremental --apply`
to fix the format errors.

There is are are also lint errors and coverage errors.  You should be able to see those at https://travis-ci.com/quantumlib/Cirq/builds/121240226

mpharrigan commented on cirq/devices/noise_model_test.py:
```diff
@@ -1,115 +1,126 @@
-
-    with pytest.raises(ValueError, match='num_qubits'):
-        _ = cirq.ConstantQubitNoiseModel(cirq.CNOT**0.01)
+# Copyright 2019 The Cirq Developers
```
this whole file is showing up as a deletion and addition of every line. I've seen this before when different line-endings (windows vs. other) are committed. Can you check your line-ending encoding to reduce the diff here? We use unix-style line endings

vtomole commented:
Ping @DuckTigger

vtomole commented:
Pin again @DuckTigger

DuckTigger commented:
Apologies for the delay.

vtomole commented:
@DuckTigger Please fix the formatting by running `check/format-incremental --apply`

DuckTigger commented:
I've ran `check/format-incremental --apply` however it is still failing the format test. 
1) Do I need to make the recommended changes manually?
2)When I run `check/format-incremental --apply` it also changes lots of other files, do I need to commit those changes also?

vtomole commented:
@DuckTigger It worked when I ran your branch on my machine. Only ` cirq/devices/two_qubit_noise_model.py` was changed.

```
(venv3.6) vtomole@vtomole:~/Cirq$ check/format-incremental --apply
Comparing against revision 'upstream/master'.
Reformatted changed lines.
(venv3.6) vtomole@vtomole:~/Cirq$ git branch
* DuckTigger-two_qubit_noise
```

DuckTigger commented:
Okay, well I'm not sure where I'm going wrong then, have deleted the directory and re-cloned into to test, and get the same result:
`andrew@andrew-HP:~/Documents/PhD$ rm -r -f cirq_fork/
andrew@andrew-HP:~/Documents/PhD$ git clone git@github.com:DuckTigger/Cirq cirq_fork
Cloning into 'cirq_fork'...
remote: Enumerating objects: 16, done.
remote: Counting objects: 100% (16/16), done.
remote: Compressing objects: 100% (16/16), done.
remote: Total 17072 (delta 5), reused 4 (delta 0), pack-reused 17056
Receiving objects: 100% (17072/17072), 6.93 MiB | 2.15 MiB/s, done.
Resolving deltas: 100% (13427/13427), done.
andrew@andrew-HP:~/Documents/PhD$ cd ./cirq_fork/
andrew@andrew-HP:~/Documents/PhD/cirq_fork$ git checkout two_qubit_noise
Branch 'two_qubit_noise' set up to track remote branch 'two_qubit_noise' from 'origin'.
Switched to a new branch 'two_qubit_noise'
andrew@andrew-HP:~/Documents/PhD/cirq_fork$ ./check/format-incremental --apply
Comparing against revision 'origin/master'.
Reformatted changed lines.
andrew@andrew-HP:~/Documents/PhD/cirq_fork$ git status
On branch two_qubit_noise
Your branch is up-to-date with 'origin/two_qubit_noise'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   cirq/_compat.py
	modified:   cirq/_compat_test.py
	modified:   cirq/_import.py
	modified:   cirq/aqt/aqt_device.py
	modified:   cirq/circuits/circuit.py
	modified:   cirq/circuits/circuit_dag.py
	modified:   cirq/circuits/circuit_test.py
	modified:   cirq/circuits/optimization_pass.py
	modified:   cirq/circuits/qasm_output.py
	modified:   cirq/contrib/acquaintance/bipartite.py
	modified:   cirq/contrib/acquaintance/executor.py
	modified:   cirq/contrib/acquaintance/inspection_utils.py
	modified:   cirq/contrib/acquaintance/mutation_utils.py
	modified:   cirq/contrib/acquaintance/optimizers.py
	modified:   cirq/contrib/acquaintance/permutation.py
	modified:   cirq/contrib/acquaintance/shift.py
	modified:   cirq/contrib/acquaintance/shift_swap_network.py
	modified:   cirq/contrib/acquaintance/shift_test.py
	modified:   cirq/contrib/acquaintance/strategies/complete.py
	modified:   cirq/contrib/acquaintance/strategies/cubic.py
	modified:   cirq/contrib/acquaintance/strategies/quartic_paired.py
	modified:   cirq/contrib/acquaintance/topological_sort.py
	modified:   cirq/contrib/graph_device/uniform_graph_device.py
	modified:   cirq/contrib/noise_models/noise_models.py
	modified:   cirq/contrib/qasm_import/_parser.py
	modified:   cirq/contrib/quantum_volume/quantum_volume.py
	modified:   cirq/contrib/routing/greedy.py
	modified:   cirq/contrib/routing/initialization.py
	modified:   cirq/contrib/routing/utils.py
	modified:   cirq/devices/line_qubit.py
	modified:   cirq/devices/line_qubit_test.py
	modified:   cirq/devices/noise_model.py
	modified:   cirq/experiments/cross_entropy_benchmarking.py
	modified:   cirq/experiments/fidelity_estimation.py
	modified:   cirq/experiments/google_v2_supremacy_circuit.py
	modified:   cirq/experiments/n_qubit_tomography.py
	modified:   cirq/experiments/qubit_characterizations.py
	modified:   cirq/experiments/t1_decay_experiment.py
	modified:   cirq/google/api/v1/programs.py
	modified:   cirq/google/arg_func_langs.py
	modified:   cirq/google/arg_func_langs_test.py
	modified:   cirq/google/common_serializers.py
	modified:   cirq/google/common_serializers_test.py
	modified:   cirq/google/devices/known_devices.py
	modified:   cirq/google/devices/serializable_device.py
	modified:   cirq/google/devices/xmon_device.py
	modified:   cirq/google/devices/xmon_device_test.py
	modified:   cirq/google/engine/calibration.py
	modified:   cirq/google/engine/calibration_test.py
	modified:   cirq/google/engine/engine.py
	modified:   cirq/google/engine/engine_job.py
	modified:   cirq/google/engine/engine_program.py
	modified:   cirq/google/engine/engine_test.py
	modified:   cirq/google/op_deserializer.py
	modified:   cirq/google/op_deserializer_test.py
	modified:   cirq/google/op_serializer.py
	modified:   cirq/google/op_serializer_test.py
	modified:   cirq/google/optimizers/optimize_for_sycamore.py
	modified:   cirq/google/optimizers/optimize_for_xmon.py
	modified:   cirq/google/optimizers/two_qubit_gates/gate_compilation.py
	modified:   cirq/google/optimizers/two_qubit_gates/math_utils.py
	modified:   cirq/google/serializable_gate_set.py
	modified:   cirq/google/serializable_gate_set_test.py
	modified:   cirq/interop/quirk/cells/arithmetic_cells.py
	modified:   cirq/interop/quirk/cells/cell.py
	modified:   cirq/interop/quirk/cells/composite_cell.py
	modified:   cirq/interop/quirk/cells/control_cells.py
	modified:   cirq/interop/quirk/cells/frequency_space_cells.py
	modified:   cirq/interop/quirk/cells/input_cells.py
	modified:   cirq/interop/quirk/cells/input_rotation_cells.py
	modified:   cirq/interop/quirk/cells/measurement_cells.py
	modified:   cirq/interop/quirk/cells/parse.py
	modified:   cirq/interop/quirk/cells/qubit_permutation_cells.py
	modified:   cirq/interop/quirk/cells/single_qubit_rotation_cells.py
	modified:   cirq/interop/quirk/cells/testing.py
	modified:   cirq/interop/quirk/url_to_circuit.py
	modified:   cirq/interop/quirk/url_to_circuit_test.py
	modified:   cirq/ion/ion_decomposition.py
	modified:   cirq/linalg/combinators.py
	modified:   cirq/linalg/decompositions.py
	modified:   cirq/linalg/diagonalize.py
	modified:   cirq/linalg/operator_spaces.py
	modified:   cirq/linalg/predicates.py
	modified:   cirq/ops/arithmetic_operation.py
	modified:   cirq/ops/clifford_gate.py
	modified:   cirq/ops/common_gates.py
	modified:   cirq/ops/controlled_gate.py
	modified:   cirq/ops/controlled_operation.py
	modified:   cirq/ops/dense_pauli_string.py
	modified:   cirq/ops/dense_pauli_string_test.py
	modified:   cirq/ops/eigen_gate.py
	modified:   cirq/ops/eigen_gate_test.py
	modified:   cirq/ops/gate_operation.py
	modified:   cirq/ops/identity.py
	modified:   cirq/ops/linear_combinations.py
	modified:   cirq/ops/linear_combinations_test.py
	modified:   cirq/ops/matrix_gates.py
	modified:   cirq/ops/measure_util.py
	modified:   cirq/ops/measurement_gate.py
	modified:   cirq/ops/measurement_gate_test.py
	modified:   cirq/ops/op_tree.py
	modified:   cirq/ops/parallel_gate_operation.py
	modified:   cirq/ops/parity_gates.py
	modified:   cirq/ops/pauli_gates.py
	modified:   cirq/ops/pauli_interaction_gate.py
	modified:   cirq/ops/pauli_string.py
	modified:   cirq/ops/pauli_string_phasor.py
	modified:   cirq/ops/pauli_string_test.py
	modified:   cirq/ops/phased_iswap_gate.py
	modified:   cirq/ops/phased_x_gate.py
	modified:   cirq/ops/raw_types.py
	modified:   cirq/ops/swap_gates.py
	modified:   cirq/ops/three_qubit_gates.py
	modified:   cirq/optimizers/eject_phased_paulis.py
	modified:   cirq/optimizers/merge_interactions.py
	modified:   cirq/optimizers/two_qubit_to_fsim.py
	modified:   cirq/optimizers/two_qubit_to_fsim_test.py
	modified:   cirq/protocols/apply_channel_protocol.py
	modified:   cirq/protocols/apply_mixture_protocol.py
	modified:   cirq/protocols/apply_mixture_protocol_test.py
	modified:   cirq/protocols/apply_unitary_protocol.py
	modified:   cirq/protocols/circuit_diagram_info_protocol.py
	modified:   cirq/protocols/commutes.py
	modified:   cirq/protocols/decompose_protocol.py
	modified:   cirq/protocols/decompose_protocol_test.py
	modified:   cirq/protocols/json_serialization.py
	modified:   cirq/protocols/mixture_protocol.py
	modified:   cirq/protocols/pauli_expansion_protocol.py
	modified:   cirq/protocols/pow_protocol.py
	modified:   cirq/sim/clifford/clifford_simulator.py
	modified:   cirq/sim/clifford/clifford_tableau.py
	modified:   cirq/sim/density_matrix_utils.py
	modified:   cirq/sim/mux.py
	modified:   cirq/sim/simulator.py
	modified:   cirq/sim/sparse_simulator.py
	modified:   cirq/sim/wave_function.py
	modified:   cirq/sim/wave_function_simulator.py
	modified:   cirq/study/flatten_expressions.py
	modified:   cirq/study/flatten_expressions_test.py
	modified:   cirq/study/resolver.py
	modified:   cirq/study/sweepable.py
	modified:   cirq/study/trial_result.py
	modified:   cirq/study/trial_result_test.py
	modified:   cirq/testing/circuit_compare.py
	modified:   cirq/testing/consistent_protocols.py
	modified:   cirq/testing/lin_alg_utils.py
	modified:   cirq/testing/sample_circuits.py
	modified:   cirq/value/angle.py
	modified:   cirq/value/random_state.py
	modified:   cirq/vis/heatmap.py
	modified:   cirq/work/pauli_sum_collector.py
	modified:   dev_tools/bash_scripts_test.py
	modified:   examples/advanced/quantum_volume_test.py
	modified:   examples/direct_fidelity_estimation.py
	modified:   examples/shor.py

no changes added to commit (use "git add" and/or "git commit -a")
andrew@andrew-HP:~/Documents/PhD/cirq_fork$ git branch
  master
* two_qubit_nois
`

vtomole commented:
`Comparing against revision 'origin/master'.`

This might be it. Your `origin/master` is  419 commits behind quantumlib:master.  Let's try pushing those changes your formatter made.

vtomole commented:
@DuckTigger Try syncing your fork's master branch to Cirq's master

DuckTigger commented:
Yes, this branch didn't have qunatumlib/Cirq set as upstream;  so after fixing that I ran `format-incremental` which still failed the format check. This was 4 lines, which I fixed manually and committed, then running `format-incremental` undoes the manual changes. I have pushed the previous commit

vtomole commented on cirq/devices/two_qubit_noise_model.py:
```diff
@@ -0,0 +1,840 @@
+
+# noinspection PyProtectedMember
+@cirq.value.value_equality
+class TwoQubitDepolarizingChannel(gate_features.TwoQubitGate):
```
Let's move this to `cirq/ops`

DuckTigger replied:
Does this include the asymmetric depolarising channel and all of the associated gates here?

vtomole replied:
Yes. Channels should also be in `ops`.

vtomole commented on cirq/devices/two_qubit_noise_model.py:
```diff
@@ -0,0 +1,840 @@
+            self._p_zy, self._p_zz, self._p_iz, self._p_ix, self._p_iy)
+
+
+def two_qubit_asymmetric_depolarize(p_xi: float, p_yi: float, p_zi: float,
```
Add documentation.

DuckTigger replied:
In the one qubit depolarising channel this isn't documented. I've added a short descripton. Should I also describe all parameters with `:param:`?

vtomole replied:
> Should I also describe all parameters with :param:?

Yes please.

vtomole commented on cirq/devices/two_qubit_noise_model.py:
```diff
@@ -0,0 +1,840 @@
+        self.single_qubit_noise_gate = single_qubit_noise_gate
+        self.two_qubit_noise_gate = two_qubit_noise_gate
+
+    def noisy_operation(self, operation: cirq.Operation):
```
Add return type and documentation.

DuckTigger commented on cirq/devices/two_qubit_noise_model.py:
```diff
@@ -0,0 +1,840 @@
+                args, kwargs))
+
+
+def construct_eigen_space(mat):
```
This was used to construct the eigen_components of each gate. i have left it here for transparency, but it can be removed if not helpful.

vtomole replied:
Nice! `eigen_components` is one of the most confusing parts of Cirq IMO. Let me run potentially opening a new pull request for this added to https://github.com/quantumlib/Cirq/tree/master/cirq/linalg with a blurb about it on https://github.com/quantumlib/Cirq/blob/master/docs/gates.md by @Strilanc .

@Strilanc When do `eigen_components` need to be defined for gates? https://github.com/quantumlib/Cirq/blob/f7f224c585ea7e75205144edf08d92fd2dc872b2/cirq/ops/fsim_gate.py#L37 doesn't have one.

DuckTigger replied:
The reason I defined `_eigen_components` and `_decompose` is that I believe the former is quicker?

Strilanc replied:
It's needed when you inherit EigenGate.

vtomole replied:
Remove it. We don't need it.

vtomole commented:
@DuckTigger Are you still having to fix the formating issues manually? If so,  Try updating your master branch to match this master branch, then merge your master into this branch.

DuckTigger commented:
> @DuckTigger Are you still having to fix the formating issues manually? If so, try updating your master branch to match this master branch, then merge your master into this branch.

I have pulled the remote and merged with master, but running `check\format-incremental` still makes changes which don't pass these tests

vtomole commented:
@DuckTigger  What's your OS?

DuckTigger commented:
> @DuckTigger What's your OS?

```
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Ubuntu 18.04.3 LTS"
```

DuckTigger commented:
Structure of tests copied from the corresponding single qubit channels - currently failing the `repr` tests and a few `asset_equals` tests, but the latter are failing due to 'too much' precision. Will investigate.

vtomole commented on cirq/ops/two_qubit_noise_channels.py:
```diff
@@ -0,0 +1,1139 @@
+
+
+@cirq.value.value_equality
+class TwoQubitAsymmetricDepolarizingChannel(gate_features.TwoQubitGate):
```
Move this into a file `cirq/contrib/noise_models/two_qubit_noise_models.py`. Also add a file that tests it in that `noise_models_directory`?

vtomole commented on cirq/ops/two_qubit_noise_channels.py:
```diff
@@ -0,0 +1,1139 @@
+
+
+@cirq.value.value_equality
+class TwoQubitDepolarizingChannel(gate_features.TwoQubitGate):
```
Move this into a file `cirq/contrib/noise_models/two_qubit_noise_models.py`. Also add a file that tests it in that `noise_models_directory`?

vtomole commented:
Ping @DuckTigger

balopat commented:
Friendly ping @DuckTigger - are you planning to work on this?

dabacon commented:
This seems to have stalled out (was pinged in Feb).  I'm going to close, but if there is more work on this, please reopen.
